src/main/java/com/google/codeu/data/Datastore.java

```diff
@@ -425,39 +423,68 @@ public void storeLocation(Location location)
-            if(Math.abs(locations.get(0).getLatitude()-otherLatitude)<3 && Math.abs(locations.get(0).getLongitude()-otherLongitude)<3)
-            {
-              String idString = entity.getKey().getName();
-              UUID id = UUID.fromString(idString);
-              String text = (String) entity.getProperty("text");
-              String otherUser = (String) entity.getProperty("user");
-              Location location = new Location(id, otherLatitude, otherLongitude, text, otherUser);
-              locations.add(location);
-            }
-
-          } catch (Exception e) {
-            System.err.println("Error reading message.");
-            System.err.println(entity.toString());
-            e.printStackTrace();
+    if(locations.size()==0 || locations.size()>1)//location does not exist or multiple invalid locations
+        return null;
+    else
+        return locations.get(0);
+  }
+  
+  /*return list of charity locations near a user location*/
+  public List<Location> getAllNearCharityLocations(Location location) {   
+    List<Location> locations = new ArrayList<>();
+    Query allLocations =
+       new Query("Location")
+               .setFilter(new Query.FilterPredicate("user", FilterOperator.NOT_EQUAL, location.getUser()));
+    PreparedQuery allLocationresults = datastore.prepare(allLocations);
+    for (Entity entity : allLocationresults.asIterable()) {
+      try {
+        double otherLongitude = Double.parseDouble(entity.getProperty("longitude").toString());
+        double otherLatitude = Double.parseDouble(entity.getProperty("latitude").toString());
+        if(Math.abs(location.getLatitude()-otherLatitude)<3 && Math.abs(location.getLongitude()-otherLongitude)<3)
+        {
+          String idString = entity.getKey().getName();
+          UUID id = UUID.fromString(idString);
+          String text = (String) entity.getProperty("text");
+          String otherUser = (String) entity.getProperty("user");
+          Location locationNear = new Location(id, otherLatitude, otherLongitude, text, otherUser);
+          /*get user of the location to check if it is charity type before adding it to locations*/
+          User locationUser = getUser(otherUser);
+          if(locationUser!=null && locationUser.getType()!=null && locationUser.getType()==User.CHARITY_TYPE) {
+            locations.add(locationNear);
           }
         }
+      } catch (Exception e) {
+        System.err.println("Error reading location.");
+        System.err.println(entity.toString());
+        e.printStackTrace();
+      }
     }
-    else
-    {
-      if(locations.size()>1)
-        System.out.println("Invalid: multiple locations for a charity!!!!!!!!!!!!!");
-      locations = new ArrayList<>();//if locations.size()==0, 0r >1 =>invalid: set to empty array
+    return locations;
+  }
+
+  /*return list of charity locations near a user location*/
+  public List<Location> getAllNearLocations(Location location) {
```

Comment:
I'm not sure what the difference between these two functions are? Could you update the method comment to make it more clear?

Reply:
There is also a lot of overlap between the two - could you factor out the shared code into helper methods? ie. 

private PreparedQuery getAllLocations(User user) {
   Query allLocations =
       new Query("Location")
               .setFilter(new Query.FilterPredicate("user", FilterOperator.NOT_EQUAL, location.getUser()));
    return datastore.prepare(allLocations);
}

Can also factor out 445-451 as well.

Reply:
Yes, that's a good idea. I made a helper function and factored out 45 to 51 and two more lines, in general the lines that convert an Entity to Object. I didn't factor out the query part though because the queries are different queries, one is NOT_EQUAL, and one is EQUAL.

Reply:
I just realized these two function are so similar that I can call one inside the other one!